Feature/sdk datastructures

1tylermitchell opened the pull request:
These were updated after 4.5 merge, removed "uncommitted" labels, etc.
Note all files are in sdk/<lang> except sdk/core-operations - where I added some basic intro information.

mattcarabine commented:
^ Apologies for the malformed output above, just testing something - CV looks like it passed.

mattcarabine commented:
Closing + Re-opening to retrigger CV.

mattcarabine commented:
Closing + Re-opening to retrigger CV - so that we get the list of the changed files.

mattcarabine commented:
Retriggering CV so that the pages will have the correct CSS

mattcarabine commented on content/sdk/c/datastructures.dita:
```diff
@@ -0,0 +1,177 @@
+                a context similar to the following (refer to the <xref
+                    href="document-operations.dita#topic_ltm_4n5_2v">Sub-Document Operations</xref>
+                page for more
+                details).<codeblock>lcb_SDSPEC spec = { 0 };
```
Would be good to format this with C syntax highlighting (`language-c`).

mattcarabine replied:
Worth going through the whole page and adding this wherever there is a C codeblock, won't add a comment on each one as there are quite a few.

mattcarabine commented on content/sdk/core-operations.dita:
```diff
@@ -205,6 +205,51 @@ if should_increment_value(value):
+        <p>To add an item to a map, specify
+            the <i>document ID</i> of the map itself (i.e. the ID which uniquely identifies the map
+            in the server), the key <i>within</i> the map, and the value to store under the
+            key:<codeblock>bucket.map_add('map_id', 'name', 'Mark Nunberg', create=True)</codeblock></p>
```
Could add python syntax highlighting here.

mattcarabine commented on content/sdk/core-operations.dita:
```diff
@@ -205,6 +205,51 @@ if should_increment_value(value):
+            appropriate methods. Most data access methods will return an
+                <apiname>ValueResult</apiname>-like object with the actual returned value under the
+                <varname>value</varname>
+            property.<codeblock>bucket.list_get(0).value  # 'hello'
```
Could add python syntax highlighting here.

mattcarabine commented on content/sdk/dotnet/datastructures.dita:
```diff
@@ -0,0 +1,561 @@
+                id="ul_txf_32t_5x">
+                <li><b>Map</b> is like .NET <codeph>Dictionary&lt;TKey, TValue></codeph>, and is a
+                    key-value structure, where a value is accessed by using a key string.</li>
+                <li><b>List</b> is like a .NET <codeph>List&lt;TValue></codeph>s and is a sequential
```
Might be better to rephrase as 'List is like a .NET List' (as opposed to Lists).

mattcarabine commented on content/sdk/dotnet/datastructures.dita:
```diff
@@ -0,0 +1,561 @@
+                    <apiname>ISet&lt;TValue></apiname>, <apiname>IList&lt;TValue></apiname>,
+                    <apiname>IDictionary&lt;Tkey, TValue></apiname> and a
+                    <apiname>CouchbaseQueue</apiname> class. See the section below called <i>"<b>The
+                        Couchbase.Collections Namespace</b>"</i> for more details.</note></p>
```
We should be able to put a link to the section here (so that users can click and be taken directly there). I will look up how to do this for you.

mattcarabine replied:
As an update, an example is in https://github.com/couchbase/docs-cb4/blob/master/content/sdk/nodejs/start-using-sdk.dita.

The section is given an id in https://github.com/couchbase/docs-cb4/blob/master/content/sdk/nodejs/start-using-sdk.dita#L71 and the id is then used as an anchor in the xref in https://github.com/couchbase/docs-cb4/blob/master/content/sdk/nodejs/start-using-sdk.dita#L15.

mattcarabine commented on content/sdk/dotnet/datastructures.dita:
```diff
@@ -0,0 +1,561 @@
+                                        href="../../developer-guide/cas-concurrency.dita" format="dita">CAS</xref> to ensure no two processs can retrieve the
+                                    same
+                                    item.<codeblock>item = bucket.QueuePop("a_queue").Value #=> 'job123'</codeblock>If
+                                    the queue is empty, then an
```
an -> a

mattcarabine commented on content/sdk/python/datastructures.dita:
```diff
@@ -0,0 +1,181 @@
+                                <entry><apiname>queue_pop</apiname></entry>
+                                <entry>Remove an item from the end queue and return
+                                    it.<codeblock>item = bucket.queue_pop('a_queue') #=> 'job123'</codeblock>If
+                                    the queue is empty, then an
```
an -> a

mattcarabine commented on content/sdk/dotnet/datastructures.dita:
```diff
@@ -0,0 +1,561 @@
+        </p>
+        <section>
+            <title>Concurrency</title>
+            <p>Where apprioriate, all of operations on the <apiname>IBucket</apiname> level and the
```
apprioriate -> appropriate

mattcarabine commented on content/sdk/go/datastructures.dita:
```diff
@@ -0,0 +1,172 @@
+                    server, allowing for less network overhead and lower latency.</li>
+            </ul></p>
+        <section><title>Creating a Data Structure</title>Data structures can be created implicitly
+            by specifying true for the createMap option of the operation. For example, to add an
```
Will want to wrap 'true' and 'createMap' in `<codeph>` or something similar so that they are distinct.

mattcarabine commented on content/sdk/nodejs/datastructures.dita:
```diff
@@ -0,0 +1,174 @@
+            <i>will not</i> overwrite an existing document with the same name if it already exists.
+            It is therefore always safe to use, unless your application creates data structures
+                independently.<p>Data structures can be explicitly created and reset using
+                full-document methods, and initialize it with its JSON equivalent. To create a new
```
initialize it -> initialized

mattcarabine commented on content/sdk/java/datastructures.dita:
```diff
@@ -0,0 +1,179 @@
+            already exists it will not be overwritten, more elements will be pushed to the list.
+            It is therefore always safe to use, unless your application creates data structures
+                independently.<p>Data structures can be explicitly created and reset using
+                full-document methods, and initialize it with its JSON equivalent. To create a new
```
initialize it -> initialized

mattcarabine commented on content/sdk/php/datastructures.dita:
```diff
@@ -0,0 +1,17 @@
+        structures may be manipulated with basic operations without retrieving and storing the
+        entire document.</shortdesc>
+    <body>
+        <p>Data structures in Couchbase are currently not yet
```
I would say 'currently not implemented' or 'not yet implemented' here.

mattcarabine commented on content/sdk/python/datastructures.dita:
```diff
@@ -0,0 +1,181 @@
+                <li><b>Map</b> is like Python <codeph>dict</codeph>, and is a key-value structure,
+                    where a value is accessed by using a key string.</li>
+                <li><b>List</b> is like a Python <codeph>list</codeph> and is a sequential data
+                    structure. Values can be placed in the beginning or end of a list, and can be
```
As with the other list descriptions, should probably be updated to clarify.

1tylermitchell replied:
as intended

mattcarabine commented on content/sdk/python/datastructures.dita:
```diff
@@ -0,0 +1,181 @@
+def map_set(bucket, docid, key, value):
+    bucket.mutate_in(docid, SD.upsert(key, value)</codeblock>Some
+            data structure operations are compound operations. This means that they do not
+            correspond to a single server-side operation. For example, <apiname>set_remove</apiname>
```
set_remove -> the operation set_remove

MikeGoldsmith commented:
@mattcarabine I've tried to use an xref link like you suggested for dotnet/datastructures.dita, but the output file doesn't seem to work. Can you check what I've done please?

mattcarabine commented:
Sure @MikeGoldsmith, will take a look.

mattcarabine commented on content/sdk/dotnet/datastructures.dita:
```diff
@@ -0,0 +1,563 @@
+                wrapper around the various key-value APIs, you are free to switch between them in
+                your code.</note></section>
+        <section>
+            <title id="couchbase-collections">The Couchbase.Collections Namespace</title>
```
@MikeGoldsmith The reason that the anchor doesn't work properly is because the 'id' needs to be associated with the section, not the title. Moving the  `id="couchbase-collections"` into the section tag should make the xref properly.

mattcarabine commented:
Think there might be an issue with the builder, ignore the failure for now

1tylermitchell commented on content/sdk/nodejs/datastructures.dita:
```diff
@@ -0,0 +1,174 @@
+                <li><b>Map</b> is like a JavaScript <codeph>Object</codeph>, and is a key-value
+                    structure, where a value is accessed by using a key string.</li>
+                <li><b>List</b> is like a JavaScript <codeph>array</codeph>s and is a sequential
+                    data structure. Values can be placed in the beginning or end of a list, and can
```
written as intended

1tylermitchell commented:
@mattcarabine I've addressed all comments and made edits throughout.  
There was a merge conflict I manually fixed, but now I see it was a "merge branch 'master' into the feature branch?  Confused there, I guess I broke something.  Can you have a peek?

1tylermitchell commented:
FYI - PHP docs will be added to this PR very soon by @avsej  - so we aren't 100% done yet

mattcarabine commented:
@cb-docs-robot rebuild

mattcarabine commented:
Due to some weird merge issues, moving to #1053
